src/Simulation/Simulators/SparseSimulator/Native/SparseSimulator.h

```diff
@@ -0,0 +1,846 @@
+
+		// Returns the probability of a given measurement in a Pauli basis
+		// by decomposing each pair of computational basis states into eigenvectors
+		// and adding the coefficients of the respective components
+		double MeasurementProbability(std::vector<Gates::Basis> const& axes, std::vector<logical_qubit_id> const& qubits) {
+			_execute_queued_ops(qubits, OP::Ry);
+			return _quantum_state->MeasurementProbability(axes, qubits);
+		}
+
+
+
+		bool Measure(std::vector<Gates::Basis> const& axes, std::vector<logical_qubit_id> const& qubits){
+			_execute_queued_ops(qubits, OP::Ry);
+			bool result = _quantum_state->Measure(axes, qubits);
+			// Switch basis to save space
+			// Idea being that, e.g., HH = I, but if we know 
+			// that the qubit is in the X-basis, we can apply H
+			// and execute, and this will send that qubit to all ones
+			// or all zeros; then we leave the second H in the queue
+			// Ideally we would also do that with Y, but HS would force execution,
+			// rendering it pointless
+			std::vector<logical_qubit_id> measurements;
+			for (int i =0; i < axes.size(); i++){
+				if (axes[i]==Gates::Basis::PauliX){
+					H(qubits[i]);
+					measurements.push_back(qubits[i]);
+				} 
+			}
+			_execute_queued_ops(measurements, OP::H);
+			// These operations undo the previous operations, but they will be 
+			// queued 
+			for (int i =0; i < axes.size(); i++){
+				if (axes[i]==Gates::Basis::PauliX){
+					H(qubits[i]);
+				} 
+			}
+			return result;
+		}
+
+		// Returns the amplitude of a given bitstring
+		amplitude probe(std::string label) {
+			_execute_queued_ops();
+			return _quantum_state->probe(label);
+		}
+
+		std::string Sample() {
+			_execute_queued_ops();
+			return _quantum_state->Sample();
+		}
+
+		// Dumps the state of a subspace of particular qubits, if they are not entangled
+		// This requires it to detect if the subspace is entangled, construct a new 
+		// projected wavefunction, then call the `callback` function on each state.
+		bool dump_qubits(std::vector<logical_qubit_id> qubits, void (*callback)(char*, double, double)) {
+			_execute_queued_ops(qubits, OP::Ry);
+			return _quantum_state->dump_qubits(qubits, callback);
+		}
+
+		// Dumps all the states in superposition via a callback function
+		void dump_all(logical_qubit_id max_qubit_id, void (*callback)(char*, double, double)) {
+			_execute_queued_ops();
+			_quantum_state->dump_all(max_qubit_id, callback);
+		}
+
+		// Updates state to all queued gates
+		void update_state() {
+			_execute_queued_ops();
+		}
+
+
+    private:
+
+    	// These indicate whether there are any H, Rx, or Ry gates
+    	// that have yet to be applied to the wavefunction. 
+    	// Since HH=I and Rx(theta_1)Rx(theta_2) = Rx(theta_1+theta_2)
+    	// it only needs a boolean to track them.
+    	std::vector<bool> _H_queue;
+    	std::vector<bool> _Rx_queue;
+    	std::vector<bool> _Ry_queue;
+
+    	std::vector<double> _Rx_angles;
+    	std::vector<double> _Ry_angles;
```

Comment:
FYI: The identifiers whose names start with underscore `_` followed by a capital letter (e.g. the identifier starting with `_H`) are reserved by the C++ standard. You might want ot rename such identifiers.

Reply:
Indeed. https://en.cppreference.com/w/cpp/language/identifiers

Reply:
Addressed here: https://github.com/sam-jaques/qsharp-runtime/pull/2